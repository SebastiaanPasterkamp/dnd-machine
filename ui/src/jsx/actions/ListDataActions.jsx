import React from 'react';
import Reflux from 'reflux';
import _ from 'lodash';

import ReportingActions from './ReportingActions.jsx';

function jsonOrBust(response) {
    if (response.ok) {
        return response.json();
    }
    throw response.statusText;
};

var ListDataActions = Reflux.createActions({
    "setState": {},
    "doLogin": {children: ['completed', 'failed']},
    "doLogout": {children: ['completed', 'failed']},
    "fetchItems": {children: ['completed', 'failed']},
});
let throttledGet = {};

ListDataActions.fetchItems.listen((type, category=null) => {
    let path = '/' + _.filter([
        category, type, category ? 'api' : null]).join('/');

    if (!(path in throttledGet)) {
        throttledGet[path] = _.throttle((path, type) => {

            fetch(path, {
                credentials: 'same-origin',
                'headers': {
                    'X-Requested-With': 'XMLHttpRequest'
                }
            })
            .then(jsonOrBust)
            .then((response) => {
                ListDataActions.fetchItems.completed({
                    [type]: response
                });
            })
            .catch((error) => {
                console.log(error);
                ListDataActions.fetchItems.failed(type, error);
            });
        }, 1000, {leading: true, trailing: false});
    }

    throttledGet[path](path, type);
});

ListDataActions.doLogin.listen((credentials, success, failure) => {
    fetch('/login', {
        credentials: 'same-origin',
        method: 'POST',
        'headers': {
            'X-Requested-With': 'XMLHttpRequest',
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(credentials)
    })
    .then(jsonOrBust)
    .then((response) => {
        ListDataActions.doLogin.completed(response);
        ListDataActions.fetchItems.completed({
            current_user: response
        });

        ReportingActions.showMessage(
            'good',
            'Login successful',
            'Login',
            5
        );

        if (success) {
            success(content);
        }
    })
    .catch((error) => {
        console.log(error);

        ReportingActions.showMessage(
            'bad',
            'Login failed',
            'Login',
            10
        );

        ListDataActions.doLogin.failed(error);

        if (failure) {
            failure();
        }
    });
});

ListDataActions.doLogout.listen((callback) => {
    fetch('/logout', {
        credentials: 'same-origin',
        'headers': {
            'X-Requested-With': 'XMLHttpRequest'
        }
    })
    .then(jsonOrBust)
    .then((response) => {
        ListDataActions.doLogout.completed(response);
        ListDataActions.fetchItems.completed({
            current_user: null
        });

        ReportingActions.showMessage(
            'good',
            'Logout successful',
            'Logout',
            5
        );

        if (callback) {
            callback();
        }
    })
    .catch((error) => {
        console.log(error);
        ListDataActions.doLogout.failed(error);
        ListDataActions.fetchItems.completed({
            current_user: null
        });

        ReportingActions.showMessage(
            'bad',
            'Logout failed',
            'Logout',
            10
        );

        if (callback) {
            callback();
        }
    });
});

export default ListDataActions;
