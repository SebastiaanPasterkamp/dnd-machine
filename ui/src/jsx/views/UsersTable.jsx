import React from 'react';
import PropTypes from 'prop-types';
import {
    filter,
    isEmpty,
    map,
} from 'lodash/fp';
import {Link} from 'react-router-dom';

import '../../sass/_users-table.scss';

import ListDataWrapper from '../hocs/ListDataWrapper.jsx';
import ObjectDataListWrapper from '../hocs/ObjectDataListWrapper.jsx';

import LazyComponent from '../components/LazyComponent.jsx';
import ListLabel from '../components/ListLabel.jsx';
import UserLinks from '../components/UserLinks.jsx';

const UserHeader = function() {
    return (
        <thead>
            <tr>
                <th>Name</th>
                <th>Role</th>
                <th>Email</th>
            </tr>
        </thead>
    );
};

const UserFooter = function() {
    return (
        <tbody>
            <tr>
                <td colSpan={3}>
                    <UserLinks altStyle={true} />
                </td>
            </tr>
        </tbody>
    );
};

const UserRow = function({
    id, username, name, role, email, user_roles,
}) {
    return (
        <tr data-id={id}>
            <th>
                {username}<br/>
                <i>({name})</i>
                <UserLinks altStyle={true} id={id} />
            </th>
            <td className="users-table__roles">
                {map((r) => (
                    <ListLabel
                        key={r}
                        items={user_roles}
                        value={r}
                    />
                ))(role)}
            </td>
            <td>{email}</td>
        </tr>
    );
};

UserRow.propTypes = {
    id: PropTypes.number.isRequired,
    username: PropTypes.string.isRequired,
    name: PropTypes.string,
    role: PropTypes.arrayOf(PropTypes.string),
    email: PropTypes.string,
    user_roles: PropTypes.array,
};

UserRow.defaultProps = {
    name: '',
    email: '',
    user_roles: [],
};

class UserTable extends LazyComponent
{
    shouldDisplayRow(pattern, {name, username, email}) {
        return (
            (name && name.match(pattern))
            || (username && username.match(pattern))
            || (email && email.match(pattern))
        );
    }

    render() {
        const { users, user_roles, search } = this.props;

        if (isEmpty(users)) {
            return null;
        }

        const pattern = new RegExp(search, "i");
        const filtered = filter(
            (user) => this.shouldDisplayRow(pattern, user)
        )(users);

        return (
            <div>
                <h2 className="icon fa-address-book-o">Users</h2>

                <table className="nice-table users-table condensed bordered responsive">
                    <UserHeader />
                    <tbody key="tbody">
                        {map((user) => (
                            <UserRow
                                key={user.id}
                                user_roles={user_roles}
                                {...user}
                            />
                        ))(filtered)}
                    </tbody>
                    <UserFooter />
                </table>
            </div>
        );
    }
}

UserTable.propTypes = {
    users: PropTypes.object,
    user_roles: PropTypes.array,
    search: PropTypes.string,
};

UserTable.defaultProps = {
    search: '',
    user_roles: [],
    users: {},
};

export default ListDataWrapper(
    ObjectDataListWrapper(
        UserTable,
        {users: {type: 'user'}}
    ),
    [
        'search',
        'user_roles',
    ],
    'items'
);
