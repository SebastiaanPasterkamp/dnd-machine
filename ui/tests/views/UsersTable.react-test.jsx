import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import MockRouter from 'react-mock-router';

import ListDataActions from 'actions/ListDataActions';
import {
    mockedApi,
    user_roles,
} from '../__mocks__';

import UsersTable from 'views/UsersTable.jsx';

describe('UsersTable', () => {
    const current_user = {
        id: 1,
        username: "yankees",
        name: "Jan Kees",
        role: [ "admin", "player", "dm" ],
        email: "foo@bar.com",
    };
    const fullProps = {
        current_user,
        users: {
            [current_user.id]: current_user,
        },
    };

    beforeAll(() => {
        fetch.mockImplementation( mockedApi({
            current_user,
            user_roles,
            'user/api': fullProps.users,
        }) );
        ListDataActions.fetchItems('current_user');
        ListDataActions.fetchItems('user_roles', 'items');
    })

    afterAll(() => {
        fetch.resetMocks()
    })

    describe('render', () => {
        it('should work without props', () => {
            const tree = renderer.create(
                <MockRouter location={{pathname: `/user/list`}}>
                    <UsersTable />
                </MockRouter>
            );

            expect(tree).toMatchSnapshot();
        });

        it('should work with all props', () => {
            const tree = renderer.create(
                <MockRouter location={{pathname: `/user/list`}}>
                    <UsersTable
                        {...fullProps}
                    />
                </MockRouter>
            );

            expect(tree).toMatchSnapshot();
        });

        it('should filter', () => {
            const wrapper = mount(
                <MockRouter location={{pathname: `/user/list`}}>
                    <UsersTable
                        {...fullProps}
                        search="john"
                    />
                </MockRouter>
            );

            expect(wrapper
                .find('.users-table .base-link-group')
                .length
            ).toBe(1);
        });
    });
});
