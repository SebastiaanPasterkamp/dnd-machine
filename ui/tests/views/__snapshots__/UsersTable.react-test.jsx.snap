// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersTable render should work with all props 1`] = `
<div>
  <h2
    className="icon fa-address-book-o"
  >
    Users
  </h2>
  <table
    className="nice-table users-table condensed bordered responsive"
  >
    <thead>
      <tr>
        <th>
          Name
        </th>
        <th>
          Role
        </th>
        <th>
          Email
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        data-id={1}
      >
        <th>
          yankees
          <br />
          <i>
            (
            Jan Kees
            )
          </i>
          <div
            className="base-link-group nice-btn-group"
          >
            <a
              className="cursor-pointer nice-btn-alt icon fa-eye"
              href={undefined}
              onClick={[Function]}
            >
              View
            </a>
            <a
              className="cursor-pointer nice-btn-alt icon fa-cogs"
              href="/user/raw/1"
            >
              Raw
            </a>
            <a
              className="cursor-pointer nice-btn-alt icon fa-pencil"
              href={undefined}
              onClick={[Function]}
            >
              Edit
            </a>
          </div>
        </th>
        <td
          className="users-table__roles"
        >
          <div
            className="list-label inline"
          >
            Site Administrator
          </div>
          <div
            className="list-label inline"
          >
            Player
          </div>
          <div
            className="list-label inline"
          >
            Dungeon Master
          </div>
        </td>
        <td>
          foo@bar.com
        </td>
      </tr>
    </tbody>
    <tbody>
      <tr>
        <td
          colSpan={3}
        >
          <div
            className="base-link-group nice-btn-group"
          >
            <a
              className="cursor-pointer nice-btn-alt icon fa-plus"
              href={undefined}
              onClick={[Function]}
            >
              New
            </a>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`UsersTable render should work without props 1`] = `null`;
